applications/EigenSolversApplication/custom_processes/perturb_geometry_subgrid_process.cpp

```diff
@@ -0,0 +1,190 @@
+    // Search radius. Defines the minimum distance among the nodes in the subgrid
+    double radius = minDistance;
+
+    // Get all nodes
+    ModelPart::NodesContainerType nodes = rThisModelPart.Nodes();
+    // Construct and initialize searcher classs
+    OMP_NodeSearch searcher;
+    searcher.InitializeSearch(nodes);
+
+    // Define the reduced space vector
+    std::vector<ModelPart::NodeIterator> reduced_space_nodes;
+    ResultNodesContainerType  results;
+
+    BuiltinTimer reduced_space_timer;
+    // Mark all nodes as unvisited
+    #pragma omp parallel
+    {
+        const auto it_node_begin = rThisModelPart.NodesBegin();
+        #pragma omp for
+        for (int i = 0; i < num_of_nodes; i++){
+            auto it_node = it_node_begin + i;
+            it_node->Set(VISITED,false);
+        }
+    }
+
+    // Generate reduced space
+    for (ModelPart::NodeIterator it_node =rThisModelPart.NodesBegin(); it_node != rThisModelPart.NodesEnd(); it_node++)
+    {
+        if( !it_node->Is(VISITED) ) {
+            it_node->Set(VISITED,true);
+            reduced_space_nodes.push_back(it_node);
+            results = {};
+            searcher.SearchNodesInRadiusExclusiveImplementation(nodes,it_node->GetId()-1,radius,results);
+            for( size_t i = 0; i < results.size(); i++ ){
+                results[i]->Set(VISITED,true);
+            }
+        }
+    }
+    const int num_nodes_reduced_space = reduced_space_nodes.size();
+
+    KRATOS_INFO_IF("PerturbGeometrySubgridProcess: Find Reduced Space Time", mEchoLevel > 0)
+        << reduced_space_timer.ElapsedSeconds() << std::endl
+        << "Number of Nodes in Reduced Space: " << num_nodes_reduced_space << " / " << num_of_nodes << std::endl;
+
+    // Construct and initialize correlation matrix
+    BuiltinTimer build_cl_matrix_timer;
+    Eigen::MatrixXd correlation_matrix;
+    correlation_matrix.resize(num_nodes_reduced_space,num_nodes_reduced_space);
+    // Assemble correlation matrix
+    #pragma omp parallel
+    {
+        const auto it_node_begin = reduced_space_nodes.begin();
+        #pragma omp for
+        for( int row_counter = 0; row_counter < num_nodes_reduced_space; row_counter++)
+        {
+            auto it_node = it_node_begin + row_counter;
+            for( int column_counter = 0; column_counter < num_nodes_reduced_space; column_counter++)
+            {
+                auto it_node_2 = it_node_begin + column_counter;
+                correlation_matrix(row_counter ,column_counter ) = CorrelationFunction( *it_node, *it_node_2, correlationLength);
+            }
+        }
+    }
+    KRATOS_INFO_IF("PerturbGeometrySubgridProcess: Build Correlation Matrix Time", mEchoLevel > 0)
+        << build_cl_matrix_timer.ElapsedSeconds() << std::endl;
+
+    // Construct eigensolver and solve eigenproblem
+    BuiltinTimer eigensolver_time;
+    Eigen::SelfAdjointEigenSolver<Eigen::MatrixXd> es(num_nodes_reduced_space);
```

Comment:
why don't you add this like the `EigensystemSolver`? 
Then eventually you could use it through the `EigenSolversFactory` (Eigen NOT referring to the library)

Reply:
I used the Eigen-matices in this file, because of the function argument of the eigensolver. But yes you are right, I will add it to the EigenSolversFactory and then I can wrap the Ublas matrix there..
Nevertheless, I can still not use the eigensolvers inside the StructuralMechanicsApp right? Or is there any way to do so?

Reply:
> I used the Eigen-matices in this file, because of the function argument of the eigensolver. But yes you are right, I will add it to the EigenSolversFactory and then I can wrap the Ublas matrix there..

sounds good

@loumalouomega there is no C++ EigenSolverFactory yet right? I guess this is waiting for #3185 ?

> Nevertheless, I can still not use the eigensolvers inside the StructuralMechanicsApp right? Or is there any way to do so?

you can if you use the factory (which is currently only available in Python, so you'd have to make an interface for it, just like e.g. the B&S has)

Reply:
Understand. Makes sense. Then for now we agree that I move everything to the StructuralMechanicsApp with a python interface?
Shall I open a new PR or do the changes here?

Reply:
I guess a new PR would be cleaner :)

Reply:
> Then for now we agree that I move everything to the StructuralMechanicsApp with a python interface?

yes, I guess this is the best place. Or do you have another suggestion?

Reply:
No, totally agree. Will work on that..